### ISSUE TITLE
[AIR] Sampling support for Ray Train/Ray Data

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray#31127]
### Description

For many training use cases, sampling is often important to train robust models. 

In the basic case, you want to use a RandomSampler which is equivalent to local shuffling.
For more advanced use cases, you may want weighted random sampling or stratified sampling if your training data is very imbalanced.

See PyTorch samplers for some examples: https://pytorch.org/docs/stable/data.html#torch.utils.data.Sampler

When using AIR and Ray Datasets, we should provide an API for users to implement various sampling procedures. 

Implementation wise, this would look like a windowed all-to-all operation, which may or may not be on the same process as the training worker.

**Comment by crypdick** [comment]
Related: https://github.com/ray-project/ray/issues/9337

**Comment by richardliaw** [comment]
Related to https://github.com/ray-project/ray/issues/34849

**Comment by genesis-jamin** [comment]
@anyscalesam can you point me to where stratified sampling was added?

**Comment by richardliaw** [comment]
Related slack thread here as well - https://ray.slack.com/archives/C02PHB3SQHH/p1713998236854359?thread_ts=1713997592.109249&cid=C02PHB3SQHH
